src/main/java/com/linkedin/camus/etl/kafka/coders/ProtobufMessageDecoder.java

```diff
@@ -0,0 +1,72 @@
+package com.linkedin.camus.etl.kafka.coders;
+
+import com.google.protobuf.Descriptors;
+import com.google.protobuf.Message;
+import com.google.protobuf.Timestamp;
+import com.linkedin.camus.coders.CamusWrapper;
+import com.linkedin.camus.coders.MessageDecoder;
+
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.security.InvalidParameterException;
+import java.time.Instant;
+import java.util.Properties;
+import java.util.concurrent.TimeUnit;
+
+public class ProtobufMessageDecoder extends MessageDecoder<byte[], Message> {
+    public static final String PROTOBUF_MESSAGE_CLASS_PREFIX = "camus.protobuf.message.class.";
+    public static final String CAMUS_MESSAGE_TIMESTAMP_FIELD = "camus.message.timestamp.field";
+    public static final String DEFAULT_TIMESTAMP_FIELD = "ts";
+
+    private Method messageParseMethod;
+    private String timestampFieldName;
+
+    @Override
+    public void init(Properties props, String topicName) {
+        super.init(props, topicName);
+
+        String messageClassForTopic = props.getProperty(PROTOBUF_MESSAGE_CLASS_PREFIX + topicName);
+        if (messageClassForTopic == null || messageClassForTopic.isEmpty()) {
+            throw new InvalidParameterException(PROTOBUF_MESSAGE_CLASS_PREFIX + topicName + " is required to be set");
+        }
+
+        timestampFieldName = props.getProperty(CAMUS_MESSAGE_TIMESTAMP_FIELD, DEFAULT_TIMESTAMP_FIELD);
+
+        try {
+            Class<?> eventClass = Class.forName(messageClassForTopic);
+            messageParseMethod = eventClass.getDeclaredMethod("parseFrom", byte[].class);
+        } catch (ClassNotFoundException e) {
+            throw new InvalidParameterException("Unable to load event class: " + e.getMessage());
+        } catch (NoSuchMethodException e) {
+            throw new InvalidParameterException("Unable to find parseFrom() method in protobuf message class: " + e.getMessage());
+        }
+    }
+
+    @Override
+    public CamusWrapper<Message> decode(byte[] payload) {
+        if (payload.length < 1) {
+            throw new RuntimeException("Empty payload!");
+        }
+
+        try {
+            Message msg = (Message) messageParseMethod.invoke(null, payload);
+
+            Descriptors.FieldDescriptor fd = msg.getDescriptorForType().findFieldByName(timestampFieldName);
+            if (fd != null) {
+                Object tsField = msg.getField(fd);
+                if (tsField instanceof Timestamp) {
+                    Timestamp ts = (Timestamp) tsField;
+                    return new CamusWrapper<>(msg, TimeUnit.NANOSECONDS.toMillis(ts.getNanos()));
+                }
+            }
+
+            return new CamusWrapper<>(msg, Instant.now().toEpochMilli());
```

Comment:
It would be worth having some logging here for when we fallback to the current time.

Reply:
Thanks, fixed.